matching-notification-api/src/main/java/org/phenotips/matchingnotification/storage/internal/DefaultMatchStorageManager.java

```diff
@@ -260,6 +274,72 @@ public boolean saveRemoteMatches(List<? extends PatientSimilarityView> similarit
         }
     }
 
+    private void filterMatchesForCurrentUser(List<PatientMatch> matches)
+    {
+        User currentUser = this.users.getCurrentUser();
+        String userName = currentUser.getProfileDocument().toString();
+        String userEntityListForSQL = "'" + userName + "'";
+
+        Set<Group> userGroups = this.groupManager.getGroupsForUser(currentUser);
+        for (Group g : userGroups) {
+            userEntityListForSQL += ",'" + g.getReference().toString() + "'";
+        }
+
+        try {
+            //
+            // for performance reasons, using various AccessManager-s is too slow when
+            // checking permissions for potentially 100s of patients at once. So the query below
+            // (adapted from LiveTableMacros.xml's "generateentitypermissions" macro)
+            // retrieves names of all patient documents that current user has access to, either directly
+            // or by being a collaborator, or a member of a group that is a collaborator.
+            //
+            // Since it is only used for filtering the list of matches, it does not need to do extra checks
+            // (e.g. doc.name != PatientTemplate), since those patients will never be part of a match.
+            // Similarly, no check for "matchable" or alike is done, since if a match is in the matching
+            // notification table, it will be shown as long as the user has access to the patient.
+            //
+            org.xwiki.query.Query q = this.qm.createQuery(
+                "select doc.fullName from Document as doc, "
+                + "BaseObject as obj, BaseObject accessObj, StringProperty accessProp "
+                + "where obj.name = doc.fullName and obj.className = 'PhenoTips.PatientClass' and "
+                + "accessObj.name = doc.fullName and accessProp.id.id = accessObj.id and "
+                + "((accessObj.className = 'PhenoTips.OwnerClass' and accessProp.value in "
+                + "(" + userEntityListForSQL + ")) "
+                + "or (accessObj.className = 'PhenoTips.CollaboratorClass' and accessProp.value in "
+                + "(" + userEntityListForSQL + ")) "
+                + "or (accessObj.className = 'PhenoTips.VisibilityClass' and accessProp.value in ('public', 'open')))",
+                org.xwiki.query.Query.XWQL);
+            List<String> rawDocNames = q.execute();
+
+            Set<String> patientNames = new HashSet<>();
+            for (String wikiDocName : rawDocNames) {
+                // remove the space prefix, as it is not recorded as part of a match
+                patientNames.add(wikiDocName.replaceFirst("^[^.]+\\.", ""));
+            }
+
+            // FIXME: to be removed after testing is complete. In some cases it is easier to spot errors in this
```

Comment:
Is testing complete now? You can remove this log line.

Reply:
Not yet. Need to see what happens on `preprod` first